core/lib/test_contracts/build.rs

```diff
@@ -41,22 +41,75 @@ impl ContractEntry {
     }
 }
 
+fn resolve_module_name(
+    manifest_dir: &Path,
+    test_contracts_dir: &Path,
+    factory_deps_to_include: &HashSet<String>,
+    id: &ArtifactId,
+) -> Option<String> {
+    let path_in_dir = id.source.strip_prefix(manifest_dir).ok()?;
+    let next_dir = path_in_dir.iter().next().expect("no dir");
+
+    let module_name = if next_dir.to_str() == test_contracts_dir.to_str() {
+        // We will use the test name folder and not the `contracts` folder for the module
```

Comment:
Nit: "folder" mentions here and in the comment below can be replaced to "directory" as well.